debug/kmem.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import print_function
+import gdb
+import utils
+import vars
+from ptable import ptable
+from tailq import TailQueue
+
+
+class Pool():
+    __metaclass__ = utils.GdbStructMeta
+    __ctype__ = 'kmem_pool_t'
+    __cast__ = {'mp_desc': str, 'mp_pages_used': int, 'mp_pages_max': int,
+                'mp_arena': lambda x: map(Arena,
+                                          TailQueue(field='ma_list', tq=x))}
+
+    def __init__(self, obj, name):
+        super(Pool, self).__init__(obj)
+        self.name = name
+
+
+class Arena():
+    __metaclass__ = utils.GdbStructMeta
+    __ctype__ = 'mem_arena_t'
+    __cast__ = {'ma_size': int, 'ma_flags': int, 'ma_freeblks':
+                lambda x: map(Block, TailQueue(field='mb_list', tq=x))}
+
+    def __init__(self, obj):
+        super(Arena, self).__init__(obj)
+
+
+class Block():
+    __metaclass__ = utils.GdbStructMeta
+    __ctype__ = 'mem_block_t'
+    __cast__ = {'mb_size': int, 'mb_data': str}
+
+    def __init__(self, obj):
+        super(Block, self).__init__(obj)
+
+
+class Kmem(gdb.Command, utils.OneArgAutoCompleteMixin):
+    """dump info about kernel memory pools
+
+    Syntax is: "kmem arg", where arg is one of pool names defined by
+    MALLOC_DEFINE(), e.g:
+    M_FILE, M_DEV, M_VMOBJ, M_THREAD, M_FD, M_PROC, M_VNODE, M_VFS, M_TEMP,
+    M_VMMAP, M_INITRD, M_PMAP
+
+    kmem without arguments prints a short summary of globally defined
+    pools. Locally defined pools are only accessible using "kmem arg"
+    command.
+
+    """
+
+    def __init__(self):
+        super(Kmem, self).__init__('kmem', gdb.COMMAND_USER)
+        self.pool_names = Kmem.__global_pool_names()
+
+    @staticmethod
+    def __is_local_pool_name(var):
+        return vars.has_type(var, 'kmem_pool_t *')
+
+    @staticmethod
+    def __is_global_pool_name(var):
+        return vars.has_type(var, 'kmem_pool_t *') or \
+            vars.has_type(var, 'static kmem_pool_t *')
```

Comment:
I made a quick experiment with gdb:
```
(gdb) break malloc_multiple_allocations
(gdb) n
(gdb) n
(gdb) n
(gdb) python print(gdb.parse_and_eval('M_VMOBJ').type)
kmem_pool_t *
(gdb) python print(gdb.parse_and_eval('M_TEMP').type)
kmem_pool_t *
(gdb) python print(gdb.parse_and_eval('mp').type)
kmem_pool_t *
```

`M_VMOBJ` is defined globally as static variable, `M_TEMP` is defined globally without static qualifier and `mp` is defined on stack. All three of them have the same type. Have you ever triggered a situation when gdb reports type of a variable as `static kmem_pool_t *` ?

Reply:
No. Distinction between static and non-static types was relevant in earlier commits, i.e. when pool names were read from gdb using `info variables`.  Corrected in cf3bf8d.